src/processor.cc

```diff
@@ -3,105 +3,115 @@
+	if((inputSize-nOrbitsPerDMAPacket*constants::orbit_trailer_size - 32*nOrbitsPerDMAPacket -32)%bsize!=0){
 		LOG(WARNING)
-			<< "Frame size not a multiple of block size. Will be skipped. Size="
-			<< input.size() << " - block size=" << bsize;
-		return &out;
+			<< "Frame size not a multiple of block size. Will be skipped. Frame size = "
+			<< inputSize << ", block size = " << bsize;
+		return false;
 	}
+	return true;
+}
 
+// Looks for orbit trailer then counts the non-empty bunch crossings and fills a vector with their values
+std::vector<unsigned int> StreamProcessor::CountBX(Slice& input){
+
+	char* p = input.begin() + 32; // +32 to account for orbit header
+	std::vector<unsigned int> bx_vect;
+	while( p != input.end()){
+		block1 *bl = (block1*)(p);
+		if(bl->orbit[0]==constants::beefdead){ // found orbit trailer
+			orbit_trailer *ot = (orbit_trailer*)(p);
+			for (unsigned int k = 0; k < (14*8); k++){ // 14*8 = 14 frames, 8 links of orbit trailer containing BX hitmap
+				//bxcount += __builtin_popcount(ot->bx_map[k]);
+				bit_check(&bx_vect, ot->bx_map[k], k*32);
+			}
+			return bx_vect;
+		}
 
-	while(p!=input.end()){
-		bool brill_word = false;
-		bool endoforbit = false;
-		block1 *bl = (block1*)p;
+		p+=sizeof(block1);
+	}
+
+}
+
+// Goes through orbit worth of data and fills the output memory with the muons corresponding to the non-empty bunchcrossings, as marked in bx_vect
+uint32_t StreamProcessor::FillOrbit(Slice& input, Slice& out, std::vector<unsigned int>* bx_vect){
+	char* p = input.begin() + 32; // +32 to account for orbit header
+	char* q = out.begin(); // +32 to account for orbit header
+	uint32_t relbx = 0;
+	uint32_t counts = 0;
+	while(relbx < bx_vect->size()){ //total number of non-empty BXs in orbit is given by bx_vect.size()
+		block1 *bl = (block1*)(p);
+		if(bl->orbit[0]==constants::beefdead){break;} // orbit trailer has been reached, end of orbit data
 		int mAcount = 0;
 		int mBcount = 0;
 		uint32_t bxmatch=0;
 		uint32_t orbitmatch=0;
-		uint32_t brill_marker = 0xFF;
 		bool AblocksOn[8];
 		bool BblocksOn[8];
 		for(unsigned int i = 0; i < 8; i++){
-			if(bl->orbit[i]==constants::deadbeef){
-				p += constants::orbit_trailer_size;
-				endoforbit = true;
-				break;
-			}
-			bool brill_enabled = 0;
-
-			if(( brill_enabled) && ((bl->orbit[i] == 0xFF) ||( bl->bx[i] == 0xFF) ||( bl->mu1f[i] == 0xFF) || 
-					(bl->mu1s[i] == 0xFF) ||( bl->mu2f[i] == 0xFF) ||( bl->mu2s[i] == 0xFF))){
-				brill_word = true;
-			}
-
-			//	std::cout << bl->orbit[i] << std::endl;
-			/*			if (bl->orbit[i] > 258745337){
-						std::cout << bl->orbit[i] << std::endl;
-						brill_word = true;			
-						std::cout << "orbit " << bl->orbit[i] << std::endl;
-						std::cout << "bx " << bl->bx[i] << std::endl;
-						}
-						*/
-			uint32_t bx = (bl->bx[i] >> shifts::bx) & masks::bx;
+			uint32_t bxA = (bl->bx[i] >> shifts::bx) & masks::bx;
+			uint32_t bx = bx_vect->at(relbx);
```

Comment:
if passes by a reference, then this will become `uint32_t bx = bx_vect[relbx];`, and the remaining occurrences can be converted in a similar way

Reply:
uint32_t bx = bx_vect[relbx] 

is better than

uint32_t bx = bx_vect.at(relbx)

?

Reply:
I can also use a reference for bx_count if you prefer?

Reply:
It was just a suggestion. 

Both vec[] and vec.at() are similar. vec[] a it faster and a more native way of accessing a vector/array. On the other hand with vec.at() you benefit from bounds checking at some CPU cost.  

You can keep it as it is. Up to you to decide.

Reply:
done